cpp/dolfinx/la/Matrix.h

```diff
@@ -0,0 +1,372 @@
+// Copyright (C) 2021-2022 Garth N. Wells and Chris N. Richardson
+//
+// This file is part of DOLFINx (https://www.fenicsproject.org)
+//
+// SPDX-License-Identifier:    LGPL-3.0-or-later
+
+#pragma once
+
+#include "SparsityPattern.h"
+#include <dolfinx/graph/AdjacencyList.h>
+#include <vector>
+#include <xtensor/xtensor.hpp>
+#include <xtl/xspan.hpp>
+
+namespace dolfinx::la
+{
+
+/// Distributed sparse Matrix
+template <typename T, class Allocator = std::allocator<T>>
+class Matrix
+{
+public:
+  /// The value type
+  using value_type = T;
+
+  /// Create a distributed matrix
+  Matrix(const SparsityPattern& p, const Allocator& alloc = Allocator())
+      : _index_maps({p.index_map(0), p.column_index_map()}),
+        _data(p.num_nonzeros(), 0, alloc), _cols(p.num_nonzeros()),
+        _row_ptr(
+            _index_maps[0]->size_local() + _index_maps[0]->num_ghosts() + 1, 0)
+  {
+    // Store sparsity
+    const graph::AdjacencyList<std::int32_t>& pg = p.graph();
+    std::copy(pg.array().begin(), pg.array().end(), _cols.begin());
+    std::copy(pg.offsets().begin(), pg.offsets().end(), _row_ptr.begin());
+
+    // TODO: handle block sizes
+
+    // Precompute some data for ghost updates via MPI
+    _nbr_comm = _index_maps[0]->comm(common::IndexMap::Direction::reverse);
+
+    const std::int32_t local_size0 = _index_maps[0]->size_local();
+    const std::array local_range0 = _index_maps[0]->local_range();
+    const std::int32_t num_ghosts0 = _index_maps[0]->num_ghosts();
+    const std::vector<std::int64_t>& ghosts0 = _index_maps[0]->ghosts();
+    const std::vector<int> ghost_owners0 = _index_maps[0]->ghost_owner_rank();
+    const std::int32_t local_size1 = _index_maps[1]->size_local();
+    const std::array local_range1 = _index_maps[1]->local_range();
+    const std::vector<std::int64_t>& ghosts1 = _index_maps[1]->ghosts();
+
+    // Get ghost->owner communicator for rows
+    const auto dest_ranks = dolfinx::MPI::neighbors(_nbr_comm)[1];
+    const int num_neighbors = dest_ranks.size();
+
+    // Global-to-neigbourhood map for destination ranks
```

Comment:
Comment doesn't make clear what the map is. 'global-to-local ranks for neighbourhood'?